### ISSUE TITLE
Object spreading produces wrong type with non-literal keys

### ISSUE THREAD

**Issue by coolCucumber-cat** [microsoft/TypeScript#56431]
### 🔎 Search Terms

object spread, object rest, object spreading

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about object spreading

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.2.2#code/CYUwxgNghgTiAEYD2A7AzgF0QLngb3gG0BrEAT10xgEsUBzAXVxQFcBbAIxBngF8BYAFDJ0WKLgIlylDDXpN4VWnT7wAvPiEBIcfADkegDTaAdGbBDeQA

### 💻 Code

```ts
declare const c: { [key: string]: number }
const a: { [key: string]: string } = {
	a: '',
	...c
}

```

### 🙁 Actual behavior

It allows numbers to be in the object, when they shouldn't be allowed.

### 🙂 Expected behavior

It shouldn't allow numbers.

### Additional information about the issue

Because `c` has non-literal keys, in this case `string`, which could be any string, it doesn't care what's in it.

**Comment by fatcerberus** [comment]
Just to make the issue more concrete...

[TS Playground](https://www.typescriptlang.org/play?ts=5.2.2#code/MYewdgzgLgBARgLhgbxgbQNYFMCeToBOAlmAOYC6SYArgLZxYEwC+MAvCgLABQAkIjADswnsx6hIsAIZJUmXPijEylGIRKkW7LnxkwARADMQIfQBoevAHQ24o8eAggANlivOQpABRSAlAG4eIA)

```ts
const b: { [key: string]: number } = {
	b: 777
}
const a: { [key: string]: string } = {
	a: "foo",
	...b
}
console.log(a);
```

prints:
```ts
[LOG]: {
  "a": "foo",
  "b": 777
}
```

which is a clear type violation.

**Comment by MartinJohns** [comment]
I have a feeling that this is due to #27273.

**Comment by coolCucumber-cat** [comment]
@MartinJohns Yeah that's probably it. It's been 5 years, I wonder why it's still open.

**Comment by coolCucumber-cat** [comment]
@martinjohns It's pretty damn common and it's very obvious incorrect. There are some way more useless things that are being worked on and it's also been 5 years.

**Comment by andrewbranch** [comment]
Talked briefly about this with @ahejlsberg and decided to revisit making the resulting object type `{ [key: string]: string | number }`—that is, if you spread in an index signature, the resulting type should have an index signature whose value type is the union of all property types and index signature types seen in the target object—and see what breaks in the extended test suites.

**Comment by zoto-ff** [comment]
same here

```ts
const a: Record<string, number> = { a: 1 }

new URLSearchParams({
    x: '1',
    ...a // no error, but URLSearchParams requires Record<string, string>
})
```

**Comment by wildlyinaccurate** [comment]
I stumbled upon this bug today. It really threw me off. I would expect this code to generate a type error, but it does not:

```ts
type MyType = {
	x: string;
};

const bad = { y: 123 };

const result = {
    x: "hello",
    ...bad,
} satisfies MyType;
```

**Comment by RyanCavanaugh** [comment]
@wildlyinaccurate that's unrelated and is working as intended. The object *does* meet the tested type

**Comment by wildlyinaccurate** [comment]
> [@wildlyinaccurate](https://github.com/wildlyinaccurate) that's unrelated and is working as intended. The object _does_ meet the tested type

@RyanCavanaugh sorry if I'm being dumb here, but how come manually adding the `y` property generates a type error (as expected) but using the spread operator does not? https://www.typescriptlang.org/play/?#code/C4TwDgpgBAsiAq5oF4oG8BQBIAHgLigGdgAnASwDsBzAbgwF86MBjAewuKgCMBDAEyio0UEAQCMAJgDMURhhbtOJCIQCuAG2BjB6DFH1R8UAEQALCOvWtjAGj0GAdE95879Ij2BlCAMzIrYBCQmNg5gKGU1TQkdTANDAjMLK1t7fVEoSSk3Dy9ff0JAxEgaIA

```ts
type MyType = {
	x: string;
};

const bad = { y: 123 };

const result1 = {
    x: "hello",
    ...bad,
} satisfies MyType;

const result2 = {
    x: "hello",
    y: 123,
//  ^ Object literal may only specify known properties, and 'y' does not exist in type 'MyType'.(2353)
} satisfies MyType;
```

**Comment by RyanCavanaugh** [comment]
@wildlyinaccurate see https://github.com/Microsoft/TypeScript/wiki/FAQ#indirect-excess-properties-are-ok
